improvement(std): use a general clock even in std environment

mariocynicys opened the pull request:
but this clock defaults to `StdClock` as a sane default. The user can still provide any arbitrary clock they want to provide.
This bridges up some visible gap in std vs no-std code (namely removes the phantom typer and the concrete `clock` field)

Suggest to review each commit on its own.

mariocynicys committed:
use a general clock also for std environments

this makes it so we don't dump the clock type parameter in a phandomdata but actually use it for a clock in std enviornment
but we apply a sane default type parameter `StdClock` by default

also `Default` implementation for timed map was generalized or all env
(by removing std feature guard), given that a clock implements `Default`
as well.

mariocynicys committed:
breaking-change: unify `TimedMap::new` interface for both std and non-std

breaking-change: old TimedMap::new() is to be replaced by TimedMap::default()

mariocynicys committed:
move `new_with_map_kind` on its own impl block

since this method needs the clock C to implement Default, and it's only
available for std environments, we better not force non-std clocks to
implement default for a method they don't even have access to.

ps. we might wanna re-consider whether we actually want to feature gate
`new_with_map_kind` to std only in the first place!

onur-ozkan commented on src/map.rs:
```diff
@@ -175,34 +169,30 @@ pub struct TimedMap<C, K, V> {
-    #[cfg(feature = "std")]
-    pub fn new() -> Self {
-        Self::default()
-    }
```
I think this should remain as is. I would consider reverting the second commit that replaces all new methods with default. `new` API is widely preferred over `default` in the ecosystem.

mariocynicys replied:
agreed. i wanted to make the std env able to provide a clock as well with `new(clock)`.
what about making `new()` std only and `new_with_clock(clock)` generic for both platforms?

mariocynicys replied:
i kept both `new()`s as they are (one default stdclock and the other with clock param) and added a new method for std that gives the flexibility to set the clock and the mapkind:
https://github.com/ozmann-industries/timed-map/pull/2/commits/1536cc559553d65989eda6d20ac6a7aaf7702291

onur-ozkan replied:
As we no longer need to define clock type on std platforms, we can remove re-exporting `StdClock` type, and also remove the flexibility of providing custom clocks on std platforms (https://github.com/ozmann-industries/timed-map/commit/1536cc559553d65989eda6d20ac6a7aaf7702291) because it doesn't make sense anymore.

LGTM other than that.

mariocynicys replied:
> As we no longer need to define clock type on std platforms, we can remove re-exporting StdClock type

aha agreed. to stop exporting `StdClock`, but keeping `Clock` exported? right?

> and also remove the flexibility of providing custom clocks on std platforms (https://github.com/ozmann-industries/timed-map/commit/1536cc559553d65989eda6d20ac6a7aaf7702291) because it doesn't make sense anymore.

but then an std guy can't create `CustomClock: impl Clock` and provide it! Not that i want this feat specifically, but thinking it's a good thing to have. (for `CustomClock: impl Clock` to work, `Clock` trait needs to be exported).

onur-ozkan replied:
> aha agreed. to stop exporting StdClock, but keeping Clock exported? right?

for no-std platforms, yes, `Clock` needs to be exported.

> but then an std guy can't create CustomClock: impl Clock and provide it! Not that i want this feat specifically, but thinking it's a good thing to have. (for CustomClock: impl Clock to work, Clock trait needs to be exported).

On a no-std environment, you don't have a built-in clock system provided from Rust, because you don't have an environment to fetch the time data. Instead, you abstact certain hardware ticks and try to simulate a clock to measure elapsed time between 2 points (there is no standard way of doing this because each MCU would have entirely different hardware interface). With `Clock` trait, you can make that simulated clock implementation to work with timed-map expirations, I can't find a realistic scenario where std guy would need this trait. Please do let me know if you have an example case for this trait, otherwise we don't need to complicate the std API with that.

mariocynicys replied:
> Please do let me know if you have an example case for this trait, otherwise we don't need to complicate the std API with that.

i have no use case in mind tbh.
removed the exports & reverted https://github.com/ozmann-industries/timed-map/commit/1536cc559553d65989eda6d20ac6a7aaf7702291. keeping it simple till a use case arise :)

onur-ozkan commented on src/map.rs:
```diff
@@ -175,34 +169,30 @@ pub struct TimedMap<C, K, V> {
-    }
-
     /// Creates an empty map based on the chosen map implementation specified by `MapKind`.
-    #[cfg(feature = "std")]
     pub fn new_with_map_kind(map_kind: MapKind) -> Self {
```
From https://github.com/ozmann-industries/timed-map/pull/2/commits/e6ee6b9d9402875ba7a7e70b8bcca25c47b40467:

> ps. we might wanna re-consider whether we actually want to feature gate
`new_with_map_kind` to std only in the first place!

This won't work on no-std platforms.

mariocynicys replied:
aha i see now nostd has no hashmaps

onur-ozkan reviewed (commented):
neat work:) weird that CI didn't trigger..

mariocynicys committed:
Revert "breaking-change: unify `TimedMap::new` interface for both std and non-std"

This reverts commit 2b9941299ee6edf1461741bd28ccb2f4c5e627c5.

mariocynicys committed:
specialize ::new() & ::new_with_map_kind to StdClocks only

this makes the type inferences smarter (not needed at all)

we probably need another new method that will injest a clock, otherwise std users don't really have much freedom in clock choice (except the unfriendly Default impl)

mariocynicys committed:
remove more type annotation from nostd

it's smart enough to know the type since new(clock) takes a clock of known type

mariocynicys committed:
add new_with_clock_and_map_kind() method for the sake of completeness

mariocynicys committed:
re-export `Clock` trait for std environment

otherwise the map was never configurable xD

mariocynicys committed:
Revert "add new_with_clock_and_map_kind() method for the sake of completeness"

This reverts commit 1536cc559553d65989eda6d20ac6a7aaf7702291.

mariocynicys committed:
don't export Clock trait nor StdClock

onur-ozkan reviewed (approved):
Thanks!

onur-ozkan commented:
Lol this failed lots of CI stuff. I hate that I am unable to run CI from fork PRs! I will look into that later during the next release.
